### ISSUE TITLE
"remoteEnv" in devcontainer.json not working while attaching to container (remote container)

### ISSUE THREAD

**Issue by guoquan** [microsoft/vscode-remote-release#3030]
- VSCode Version: 1.45.1
- Local OS Version: macOS 10.15.4
- Remote OS Version: CentOS Linux 7 (Core)
- Remote Extension/Connection Type: SSH + Docker

Steps to Reproduce:

1. setup docker host in `settings.json`
```json
{
    "docker.host":"tcp://<remote>:2375"
}
```
2. login to remote in a separated terminal
```bash
ssh user@remote
```
3. start a dummy container
```bash
docker run -it --rm ubuntu /bin/sh
```
log:
```
Unable to find image 'ubuntu:latest' locally
latest: Pulling from library/ubuntu
d51af753c3d3: Pull complete
fc878cd0a91c: Pull complete
6154df8ff988: Pull complete
fee5db0ff82f: Pull complete
Digest: sha256:747d2dbbaaee995098c9792d99bd333c6783ce56150d1b11e333bbceed5c54d7
Status: Downloaded newer image for ubuntu:latest
#
```
4. start VSCode, in the "Remote Explorer - Containers", find the new container and attach to it

5. In the window for the container, open Command Palette (`⇧⌘P`) and choose `Remote-Containers: Open Container Configuration File`, a `devcontainer.json` like configuration file named after the image (`ubuntu.json` in my case) will be open.

6. Add "remoteEnv" (taken from the example in document) to the configuration
```json
{
	"remoteEnv": { "PATH": "${containerEnv:PATH}:/some/other/path" }
}
```

7. Reload the container window.

8. Hoping `PATH` is updated, try print it in a VS Code terminal
```bash
echo $PATH
```
However, I got output with no "/some/other/path":
```
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```
Log from "Dev Container"
```
[0 ms] Start: Resolving remote
[6 ms] Setting up container /nifty_moser

[15 ms] Start: Run: docker inspect --type container /nifty_moser
[195 ms] Start: Check Docker is running
[195 ms] Start: Run: docker info
[913 ms] Start: Inspecting container
[913 ms] Start: Run: docker inspect --type container b0fc3f68ad548306a5fa192c4995eebc1b2525a57eadcba55384b1e816c97d67
[975 ms] Start: Run: docker exec -i -u root -e VSCODE_REMOTE_CONTAINERS_SESSION=a4a52e9a-e0f8-4db9-b8f9-6016c66b8e4c1590086250894 b0fc3f68ad548306a5fa192c4995eebc1b2525a57eadcba55384b1e816c97d67 /bin/sh
[979 ms] Start: Run in container: uname -m
[1169 ms] x86_64
[1169 ms] 
[1170 ms] Start: Run in container: (cat /etc/os-release || cat /usr/lib/os-release) 2>/dev/null
[1182 ms] NAME="Ubuntu"
VERSION="20.04 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
[1182 ms] 
[1182 ms] Start: Run in container: cat /etc/passwd
[1188 ms] Start: Run in container: test -d /root/.vscode-server
[1192 ms] 
[1192 ms] 
[1192 ms] Start: Run in container: set -o noclobber ; mkdir -p '/root/.vscode-server/data/Machine' && { > '/root/.vscode-server/data/Machine/.writeMachineSettingsMarker' ; } 2> /dev/null
[1198 ms] 
[1198 ms] 
[1198 ms] Exit code 2
[1198 ms] Start: Run in container: test -d /root/.vscode-server/bin/5763d909d5f12fe19f215cbfdd29a91c0fa9208a
[1201 ms] 
[1201 ms] 
[1201 ms] Start: Launching Remote-Containers helper.
[1204 ms] Start: Run in container: cat <<'EOF-/tmp/vscode-remote-containers-702a68fb1e168a9077e84598253209d912034890.js' >/tmp/vscode-remote-containers-702a68fb1e168a9077e84598253209d912034890.js
[1211 ms] 
[1212 ms] 
[1212 ms] Start: Run in container: # Copy .gitconfig
[1215 ms] git not found
[1215 ms] 
[1215 ms] Start: Run in container: cat <<'EOF-/tmp/vscode-remote-containers-server-702a68fb1e168a9077e84598253209d912034890.js' >/tmp/vscode-remote-containers-server-702a68fb1e168a9077e84598253209d912034890.js
[1223 ms] 
[1224 ms] 
[1224 ms] Start: Run in container: command -v git >/dev/null 2>&1 && git config --global credential.helper '!f() { /root/.vscode-server/bin/5763d909d5f12fe19f215cbfdd29a91c0fa9208a/node /tmp/vscode-remote-containers-702a68fb1e168a9077e84598253209d912034890.js $*; }; f' || true
[1233 ms] 
[1233 ms] 
[1233 ms] Start: Run in container: set -o noclobber ; mkdir -p '/root/.vscode-server/data/Machine' && { > '/root/.vscode-server/data/Machine/.installExtensionsMarker' ; } 2> /dev/null
[1243 ms] 
[1243 ms] 
[1243 ms] Exit code 2
[1244 ms] Start: Run in container: for pid in `cd /proc && ls -d [0-9]*`; do { echo $pid ; readlink -f /proc/$pid/cwd ; xargs -0 < /proc/$pid/environ ; xargs -0 < /proc/$pid/cmdline ; } ; echo ; done 2>/dev/null
[4387 ms] Extension host agent is already running.
[4387 ms] Start: Run in container: cat /root/.vscode-server/bin/5763d909d5f12fe19f215cbfdd29a91c0fa9208a/.devport 2>/dev/null
[4417 ms] 44519
[4418 ms] 
[4421 ms] Forwarding local port 57741 to container port 44519
[4478 ms] Start: Run: docker exec -i -u root -w /root/.vscode-server/extensions b0fc3f68ad548306a5fa192c4995eebc1b2525a57eadcba55384b1e816c97d67 /bin/sh -c # Watch installed extensions
[4478 ms] Start: Run: docker exec -i -u root -w /root/.vscode-server/data/Machine b0fc3f68ad548306a5fa192c4995eebc1b2525a57eadcba55384b1e816c97d67 /bin/sh -c # Watch machine settings
[4527 ms] Start: Run: docker exec -i -u root -e VSCODE_REMOTE_CONTAINERS_SESSION=a4a52e9a-e0f8-4db9-b8f9-6016c66b8e4c1590086250894 b0fc3f68ad548306a5fa192c4995eebc1b2525a57eadcba55384b1e816c97d67 /root/.vscode-server/bin/5763d909d5f12fe19f215cbfdd29a91c0fa9208a/node -e 
[4544 ms] Start: Run: docker exec -i -u root -e VSCODE_REMOTE_CONTAINERS_SESSION=a4a52e9a-e0f8-4db9-b8f9-6016c66b8e4c1590086250894 b0fc3f68ad548306a5fa192c4995eebc1b2525a57eadcba55384b1e816c97d67 /root/.vscode-server/bin/5763d909d5f12fe19f215cbfdd29a91c0fa9208a/node -e 
```

**I test this same procedure with local containers (without step 1 and 2), it is also not working.**

I also try to add "remoteUssr"
```
{
	"remoteUser": "non_exist_user",
	"remoteEnv": { "PATH": "${containerEnv:PATH}:/some/other/path" }
}
```
and I get an error as expected, meaning I have the correct configuration file.
```
[1 ms] Start: Resolving remote
[13 ms] Setting up container /nifty_moser

[13 ms] Start: Run: docker inspect --type container /nifty_moser
[259 ms] Start: Check Docker is running
[260 ms] Start: Run: docker info
[1376 ms] Start: Inspecting container
[1376 ms] Start: Run: docker inspect --type container b0fc3f68ad548306a5fa192c4995eebc1b2525a57eadcba55384b1e816c97d67
[1468 ms] Start: Run: docker exec -i -u non_exist_user -e VSCODE_REMOTE_CONTAINERS_SESSION=f2ce3a80-875d-473a-bd6b-b95a9e279fdd1590086913161 b0fc3f68ad548306a5fa192c4995eebc1b2525a57eadcba55384b1e816c97d67 /bin/sh
[1480 ms] Start: Run in container: uname -m
[2226 ms] Start: Run in container: cat /etc/passwd
[2226 ms] Stdin closed!
[2228 ms] Shell server terminated (code: 126, signal: null)
unable to find user non_exist_user: no matching entries in passwd file
```

Does this issue occur when you try this locally?: Yes (w/ container anyway)
Does this issue occur when you try this locally and all extensions are disabled?: N/A (issue about the extension)

**Comment by guoquan** [comment]
Might be related to #2152 however, that one is specific to just "PATH" and with containers run by VS Code, not attaching to an "other" container.

**Comment by guoquan** [comment]
As far as I know, docker API (v1.25+) has an option `--env` or `-e` for command `docker exec` to set environment variables.
Should the extension make use of this option to update `remoteEnv`?

**Comment by chrmarti** [comment]
Which version of Remote-Containers is this?

**Comment by guoquan** [comment]
It is `ms-vscode-remote.remote-containers` `0.117.1`

I had `ms-azuretools.vscode-docker` `1.2.1` installed though. But I also tried with it disabled and it turns out nothing change.

**Comment by chrmarti** [comment]
Tried to reproduce with a CentOS 7 image without luck.

Does it work for other variable names than PATH?

**Comment by guoquan** [comment]
Do you mean you can or cannot reproduce with the CentOS image?
<img width="1136" alt="Screen Shot 2020-06-03 at 16 04 49" src="https://user-images.githubusercontent.com/1118270/83683932-30399d80-a5b4-11ea-8f31-bac1de139a72.png">
I tried to attach to a local (on Mac) container of CentOS image, and another variable name, it is still not going on well.

**Comment by chrmarti** [comment]
I cannot reproduce the problem. Also tried with an Ubuntu 20.04 image (realizing CentOS 7 is your remote machine not the container).

I see the variable being passed as part of the environment of the extension host process inside the container. Can you run this in the container:
```
extHostPid=`ps ax | grep extensionHost | head -n 1 | cut -f 1 -d ' '`
cat /proc/$extHostPid/environ | tr '\0' '\n' | grep HELLO
```

I get:
```
HELLO=WORLD
```

**Comment by guoquan** [comment]
Hi,
I did not get anything by 
```
extHostPid=`ps ax | grep extensionHost | head -n 1 | cut -f 1 -d ' '`
```
`$extHostPid` is empty.
Do you mean the pid of some process like
```
  161 ?        Sl     0:02 /root/.vscode-server/bin/5763d909d5f12fe19f215cbfdd29a91c0fa9208a/node /root/.vscode-server/bin/5763d909d5f12fe19f215cbfdd29a91c0fa9208a/out/bootstrap-fork --type=extensionHost --uriTransformerPath=/root/.vscode-server/bin/5763d909d5f12fe19f215cbfdd29a91c0fa9208a/out/vs/server/uriTransformer.js --useHostProxy=
```
Then I do
```
cat /proc/161/environ | tr '\0' '\n' | grep HELLO
```
I get
```
HELLO=WORLD
```
But in this same container, either ubuntu image or centos image, `echo $HELLO` gives me empty still.

**Comment by guoquan** [comment]
Hi,
In the container, VSCode terminal, I can find out shell pid (`echo $$` -> `206` in my case).
And I also examined that `206` is a direct child process to `161`.
Then I do
```
cat /proc/206/environ | tr '\0' '\n' | grep HELLO
```
and get the null output.

**Comment by chrmarti** [comment]
Just realized: Make sure you have 'Inherit Env' enabled in the settings:
![image](https://user-images.githubusercontent.com/9205389/83840018-c62d0f80-a6fd-11ea-8013-d796db3919ec.png)

**Comment by guoquan** [comment]
I see. Turning it on, I have the Env in the terminal now. Thanks.

I think that is kinda missing in the documentation though.
As the [reference](https://code.visualstudio.com/docs/remote/containers#_devcontainerjson-reference) suggests "sub-processes like terminals" will have it overridden but I did not notice it works based on another toggle "Inherit Env". Actually I did not notice it is based on inheriting env so I was not able to find the correct place to set it up.
